explorer: Add information about NFToken NFTs and Collection

vpontis opened the pull request:
This adds NFToken support to the Solana Explorer so that users can see NFToken NFTs and Collections and do some basic /exploring/. NFToken Docs: https://nftoken.so

## Feature Overview

[Loom Video](https://www.loom.com/share/362e24aa66ac4db198b3e014a99235cb)

<img width="1175" alt="CleanShot 2022-07-21 at 12 55 08@2x" src="https://user-images.githubusercontent.com/1319079/180270677-f02c0646-107e-4566-85b7-cd6a9f4b9eac.png">
<img width="1177" alt="CleanShot 2022-07-21 at 12 55 00@2x" src="https://user-images.githubusercontent.com/1319079/180270686-d376bcba-3477-47ef-8cb0-972ad3b8d22c.png">

## Code Overview

[Loom Video](https://www.loom.com/share/bdab68e55d73462a9222b105dd2316cb)

steveluscher commented on explorer/src/pages/AccountDetailsPage.tsx:
```diff
@@ -325,6 +345,8 @@ function InfoSection({ account }: { account: Account }) {
       />
     );
+  } else if (account.details?.owner.toBase58() === NFTOKEN_ADDRESS) {
+    return <NFTokenAccountSection account={account} />;
```
This could add up to 100KB gzipped to the explorer, since this is a new hard dependency of the account details page.

https://bundlephobia.com/package/@glow-xyz/nftoken-js@0.1.6

Generally, the architecture of ‘require all the capabilities up front, branch on if/else to select a small subset of those capabilities’ will result in unbounded growth of the Explorer. We'll end up with a distribution of often used and rarely used capabilities, despite paying the cost of downloading the sum total of all capabilities' code on every page load.

vpontis replied:
@steveluscher appreciate that feedback. Do you have a recommendation of how we can avoid that? 

I know Next.js has a `dynamic` import which allows bundle splitting but I'm not sure if you have set something up like that in Explorer.

vpontis replied:
@steveluscher I updated this not to pull in `@glow-xyz/nftoken` by re-implementing all of the parsing and fetching in this library. Lmk if it's good to merge now 👍 :)

steveluscher commented on explorer/src/utils/nftoken.ts:
```diff
@@ -0,0 +1,125 @@
+import axios from "axios";
```
🚨 This doesn't look like it's an explicit dependency of Explorer. The only reason this probably works at all is because it's a dependency of `@metaplex/js`.

vpontis replied:
Is it fine to add it as a dependency? Since it's already in the bundle...

steveluscher commented on explorer/src/utils/nftoken.ts:
```diff
@@ -0,0 +1,125 @@
+import axios from "axios";
+import pLimit from "p-limit";
```
Same.

steveluscher commented on explorer/src/utils/nftoken-parsers.ts/isNFTokenAccount.ts:
```diff
@@ -51,11 +58,14 @@ export const parseNFTokenCollectionAccount = (
     authority: parsed.authority,
     authority_can_update: Boolean(parsed.authority_can_update),
-    metadata_url: parsed.metadata_url,
+    metadata_url: parsed.metadata_url?.replace(/\0/g, "") ?? null,
```
What's this?

vpontis replied:
It strips null characters from the string due to the way that the Buffer string is parsed. I don't have to do this with `nftoken-js` since I have a smarter parser there.

steveluscher commented on explorer/src/components/account/nftoken-hooks.tsx:
```diff
@@ -0,0 +1,39 @@
+} => {
+  const { url } = useCluster();
+  const swrKey = [collectionAddress, "getNftsInCollection"];
+  const { data, error, mutate } = useSWR(swrKey, async () => {
+    return await NftokenFetcher.getNftsInCollection({
+      collection: collectionAddress,
+      rpcUrl: url,
+    });
+  });
```
Your fetcher must be stable (same function every render) and can't close over other variables in scope. All the information it needs, you can pass through arguments!

```typescript
async function nftsInCollectionFetcher(([_, collectionAddress, url]) => {
  return await NftokenFetcher.getNftsInCollection({
    collection: collectionAddress,
    rpcUrl: url,
  });
});

export const useCollectionNfts(...) {
  /* ... */
  const {data, error, mutate} = useSWR(
    ['getNftsInCollection, collectionAddress, url],
    nftsInCollectionFetcher,
  );
  /* ... */
}
```

This way:

* SWR won't blow your result away every render because you passed a new fetcher
* Changes to `url` will _actually_ result in data being stored in your SWR cache with that URL as a key

vpontis replied:
Hmm, I'm not sure you are correct that SWR will blow away the result because you passed a new fetcher. What makes you think that?

Fetcher is stored as a `ref`:

https://github.com/vercel/swr/blob/main/core/use-swr.ts#L90

vpontis replied:
I don't think this is an issue but made it stable anyways

steveluscher commented on explorer/src/components/account/nftoken-hooks.tsx:
```diff
@@ -0,0 +1,39 @@
+import { NftokenFetcher, NftokenTypes } from "@glow-xyz/nftoken-js";
+import useSWR, { SWRResponse } from "swr";
```
I know that elsewhere in this review I left notes on how to use SWR with a stable fetcher that caches data on the basis of your _full_ inputs, but I'm also a little concerned about this being the first and only use of SWR in Explorer (ie. adding another 4K library just to cache this _one_ set of models) when you _could_ implement it just with `useEffect`/`useState`.

If you decide to continue with SWR here, can you:

* Use the `suspense: true` property of SWR, and wrap components that use this hook in `<React.Suspense>` with a sensible fallback. See `useAnchorProgram` for how we use this pattern in Explorer!
* Make the key/fetcher changes that I suggested elsewhere in this PR.

vpontis replied:
I think SWR is nice and I'd like to use it in the future if I do more contributions to the library. I find the `useFetchAccountInfo` model pretty confusing and think it would be easier to move that cacheing to SWR. 

I'll look into suspense.

vpontis replied:
Hmm using suspense makes the code a bit more awkward since then I have to pull out the scaffold for the NFT grid into a different component. What's the advantage of using suspense? We aren't doing waterfall loading.

steveluscher replied:
I can go on for hours about the benefits of suspense, but I'll stop short by just noting that with Suspense, you won't see a flash of ‘Loading...’ content if the data fetch is fast enough. With the current implementation you will _always_ cycle through `null` before getting actual data, which causes the UI to render twice and the layout to shift.

vpontis replied:
Hmm, I don't think that's true with SWR since if the key is in the SWR cache it will return `data` on first render. I did a test for this and validated that I don't see Loading. I'll record a short video.

vpontis replied:
Ah it looks like the SWR cache isn't set up to persist across renders. I forgot about this since in Glow / Luma we usually seed the cache with a server side render. I can look deeper at Suspense...

steveluscher replied:
For sure. I was never talking about subsequent render performance. This is a solution for _first_ render performance. Read more on fetch-as-you-render vs. render-then-fetch! https://17.reactjs.org/docs/concurrent-mode-suspense.html#traditional-approaches-vs-suspense

vpontis replied:
👍 will read more. We usually solve this problem with SSR.

steveluscher commented on explorer/src/components/account/NFTokenAccountHeader.tsx:
```diff
@@ -0,0 +1,113 @@
+import { NftokenTypes } from "@glow-xyz/nftoken-js";
+import "bootstrap/dist/js/bootstrap.min.js";
```
I know you copied this from MetaplexNFTHeader, but we can't possibly start to replicate this pattern. This is 21K, gzipped, of side-effectful, non ESModule JavaScript, just to implement tooltips.

Can you do without the tooltip, or just use a `title` attribute?

vpontis replied:
I don't actually know why Bootstrap was included. I removed it and I can still see the following. So I'll remove it but I can remove it from the `package.json` since it's used in `scss`. 

<img width="406" alt="CleanShot 2022-07-29 at 16 42 42@2x" src="https://user-images.githubusercontent.com/1319079/181840246-39f542b7-9053-4795-85f0-98ce1bd10b51.png">

steveluscher commented on explorer/src/components/account/NFTokenAccountHeader.tsx:
```diff
@@ -0,0 +1,113 @@
+          <h2 className="header-title ms-1 align-items-center no-overflow-with-ellipsis">
+            {metadata
+              ? metadata.name || "No collection name was found"
+              : "Loading..."}
```
When you rewrite your data fetcher to use React Suspense, you won't need to null check like this! See `useAnchorProgram`.

steveluscher commented on explorer/src/components/account/NFTokenAccountHeader.tsx:
```diff
@@ -0,0 +1,113 @@
+import { useNftokenMetadata } from "./nftoken-hooks";
+
+export function NFTokenAccountHeader({ account }: { account: Account }) {
+  const nft = parseNFTokenNFTAccount(account);
```
There's a lot of render-time parsing going on in this PR. Since these operations are expensive (lots of ferrying data into and out of buffers, and base58 encoding things over and over) consider `useMemo()` so that you only have to calculate this stuff _once_ then you render with the result over and over.

vpontis replied:
Hmm, it's really not a performance hit tho... Have you noticed any issues loading the page? The parsing is very fast.

steveluscher commented on explorer/src/pages/AccountDetailsPage.tsx:
```diff
@@ -184,13 +199,18 @@ export function AccountHeader({
     );
   }
 
+  const nftokenNFT = isNFTokenAccount(account);
+  if (nftokenNFT && account) {
```
Make `account` required in `isNFTokenAccount()` and then null check it everywhere it's used. More efficient since `isNFTokenAccount()` needs to do less work, and you don't even call it if account is `null`.

```typescript
if (account && isNFTokenAccount(account)) {
  /* ... * /
}
```

steveluscher reviewed (dismissed):
Alright! This is much better, thanks for pulling the parsing stuff out of the megalib.

I left some notes of things to change to make this even lighter. Let me know when you've had time to burn through them.

Also, if you want to put in the work to undo that regrettable inclusion of Bootstrap _just_ to display a single tooltip on the entire site, you'd be saving Explorer like 21K gzipped in which case this PR would actually be _net negative_ on code size. That would be amazing.

vpontis commented:
@steveluscher I pushed updates with the following caveats:

- I didn't remove SWR or use `{ suspense: true }` — I don't think it's an issue but I can try and refactor things if you think it is. I don't see a benefit to refactoring and it'll take some time. 
- I didn't move parsing to `useMemo` since parsing is very fast. 
- I added deps for `axios` and `p-limit` explicitly

vpontis commented:
@steveluscher updated to use Suspense.

vpontis commented:
Hi @steveluscher I think this is good to go — I addressed your feedback, let me know if there's anything else you'd like to see.

jnwng commented on explorer/src/components/account/NFTokenAccountSection.tsx:
```diff
@@ -0,0 +1,267 @@
+          </td>
+        </tr>
+        <tr>
+          <td>Number NFTs</td>
```
```suggestion
          <td>Number of NFTs</td>
```
alternatively, "Size" would work here too

jnwng commented on explorer/src/components/account/NFTokenAccountSection.tsx:
```diff
@@ -0,0 +1,267 @@
+        <tr>
+          <td>Number NFTs</td>
+          <td className="text-lg-end">
+            <Suspense fallback={<div>Loading..</div>}>
```
```suggestion
            <Suspense fallback={<div>Loading...</div>}>
```
just a nit

jnwng commented on explorer/src/utils/nftoken-parsers.ts/isNFTokenAccount.ts:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    if (
+      Buffer.from(parsed!.discriminator).toString("base64") !== "IbRbNewPP2E="
```
looks primed to be a constant instead of an arbitrary string

jnwng commented on explorer/src/utils/nftoken-parsers.ts/isNFTokenAccount.ts:
```diff
@@ -0,0 +1,83 @@
+      return null;
+    }
+    if (
+      Buffer.from(parsed.discriminator).toString("base64") !== "RQLwA3YS2fI="
```
ripe for a constant!

jnwng commented on explorer/src/utils/nftoken.ts:
```diff
@@ -0,0 +1,141 @@
+          {
+            memcmp: {
+              offset: 0,
+              bytes: bs58.encode(Buffer.from("21b45b35ec0f3f61", "hex")),
```
and here, too

jnwng reviewed (changes requested):
thanks for engaging thus far! i know its been quite a while since you've submitted it, so i appreciate the persistence. 

sharing a little bit of context on why this has been taking a while to review (besides well-deserved vacations!): outside of the features being added to Explorer that support Solana network features, we (Solana Labs, Solana Foundation) have not been able to figure out the most robust way to provide scalable support for external contributions to the explorer. we definitely recognize the value of having the information related to NFTokens showing up in the explorer, though!

one request i'd like to make before we're comfortable getting this in: **can you move all NFToken-related UI components and libraries to be colocated in a subfolder?** i would welcome an `nftoken` or `external/nftoken` folder [here](https://github.com/solana-labs/solana/tree/master/explorer/src/components/account) to keep everything cleanly organized.

this would help us:
1. easily refactor contiguous features down the line
2. quickly (and perhaps automatically) identify and triage support requests in this folder
3. provide clear ownership for you / your team in that subfolder (make sure you know what you're signing up for!)

i know that the existing state of the world is not adhering to this (i.e., the Metaplex components), but i think this small-ish change would make our team much more comfortable and establish a good precedent. hope you understand where we're coming from!

vpontis commented:
> thanks for engaging thus far! i know its been quite a while since you've submitted it, so i appreciate the persistence.
> 
> sharing a little bit of context on why this has been taking a while to review (besides well-deserved vacations!): outside of the features being added to Explorer that support Solana network features, we (Solana Labs, Solana Foundation) have not been able to figure out the most robust way to provide scalable support for external contributions to the explorer. we definitely recognize the value of having the information related to NFTokens showing up in the explorer, though!
> 
> one request i'd like to make before we're comfortable getting this in: **can you move all NFToken-related UI components and libraries to be colocated in a subfolder?** i would welcome an `nftoken` or `external/nftoken` folder [here](https://github.com/solana-labs/solana/tree/master/explorer/src/components/account?rgh-link-date=2022-08-24T21%3A59%3A21Z) to keep everything cleanly organized.
> 
> this would help us:
> 
> 1. easily refactor contiguous features down the line
> 2. quickly (and perhaps automatically) identify and triage support requests in this folder
> 3. provide clear ownership for you / your team in that subfolder (make sure you know what you're signing up for!)
> 
> i know that the existing state of the world is not adhering to this (i.e., the Metaplex components), but i think this small-ish change would make our team much more comfortable and establish a good precedent. hope you understand where we're coming from!

Thanks @jnwng for getting to this.

With your suggested changes, would the PR be merge-able?

I submitted this > 1 month ago and I've addressed all comments a couple times, so I just want to make sure that I know everything you want changed before checking this out again, fixing merge conflicts, etc.

jnwng commented:
two additional requests:

can we remove this icon? this information here tends to be informational, and while i think its a nice touch i do want to retain some consistency on how addresses in this table are displayed
![image](https://user-images.githubusercontent.com/1118534/186734478-6f2f58f7-68de-4334-b8a4-e62d01bcc562.png)

after creating this `nftoken` folder, please add a README.md with support-related instructions

> With your suggested changes, would the PR be merge-able?

yes, as soon as the merge conflicts are resolved and the PR comments have been addressed, i will verify the functionality (please provide a couple of test cases to try out ourselves!) and as long as everything remains sound we can get it merged!

vpontis commented:
Thanks I have updated w/ all of your comments.

You can test here: http://localhost:3000/address/7u1svmSJkQdvPpiFAF8A8crW5vgZVjQbS8rSgT7j3X6x

Address: 7u1svmSJkQdvPpiFAF8A8crW5vgZVjQbS8rSgT7j3X6x

Then you can click on the Collection address to see more info about the collection.

jnwng commented:
awesome, thanks for the address. was able to get everything running and things look pretty good. only thing that came up that was a little odd, each of the images fails twice with a CORS error before finally succeeding.

![image](https://user-images.githubusercontent.com/1118534/186757880-1d16f75d-9c74-49fd-af25-94adeb4f6a1b.png)

not sure if this is an issue with the images being used via giantbomb.com or if there's something up with the rendering code... is that happening for you too?

vpontis commented:
I think that can happen sometimes with images — this will be an issue with user supplied images but I would rather get this in and address that in a separate PR.

vpontis commented:
@jnwng do you think we could we get this in? I believe there are no open blocking comments and the CORS issue certainly is not a blocker. 

I would like to add more PRs to improve the Explorer but it's a bit demoralizing to work on when the iteration time is > 1 month per PR. :/

jnwng commented:
appreciate the hard work and persistence here! should automatically be deployed when this goes through CI on `master`

vpontis commented:
Awesome, thank you @jnwng! If you have any other improvements that you'd like to see related to this, let me know and I'm happy to open more PRs :).
